datastore/src/main/scala/org/allenai/datastore/Datastore.scala

```diff
@@ -38,6 +40,27 @@ import scala.collection.JavaConverters._
   * @param s3   properly authenticated S3 client.
   */
 class Datastore(val name: String, val s3: AmazonS3Client) extends Logging {
+
+  private val random = new Random
+  private def withRetries[T](activity: String, retries: Int = 10)(f: => T): T =
+    if (retries <= 0) {
+      f
+    } else {
+      val sleepTime = random.nextInt(10000) + 10000 // sleep between 10 and 20 seconds
+      // If something goes wrong, we sleep a random amount of time, to make sure that we don't slam
+      // the server, get timeouts, wait for exactly the same amount of time on all threads, and then
+      // slam the server again.
+
+      try {
+        f
+      } catch {
+        case NonFatal(e) if !e.isInstanceOf[AccessDeniedException] && !e.isInstanceOf[DoesNotExistException] =>
+          logger.warn(s"$e while $activity. $retries retries left.")
+          Thread.sleep(sleepTime)
+          withRetries(activity, retries - 1)(f)
```

Comment:
Does this tail-recurse? If not, you might want a `NOTE` at the top of the function stating that large / infinite values for `retries` will overflow the stack.